### PR TITLE
Feature/view my plot

### PR DESCRIPTION
**Pull request by emilianowearemobile**
[FARM-81]

#### 📲 What

The list of crops + bottom bar + topbar

#### 🤔 Why
		
To let farmers see their crops, in a list.
		
#### 🛠 How
		
Add a bottombar, topbar, integration with firebase and redux architecture.

#### ✅ Acceptance criteria

- [x] Manually tested
- [x] Showcase video / automation test
- [x] Rebased/merged with latest changes from dev/release and re-tested
- [x] Removed TODOs and FIXMEs

#### 🕵️‍♂️ How to test

You can run the project and check the different tabs.

### CODE REVIEW

File: android/app/google-services.json [hunk 1]
```diff
@@ -0,0 +1,69 @@
+{
```
Comment (maribel4amido) [inline android/app/google-services.json:1]: This file is sensitive and should be excluded.  Please force so that history removes this file

Comment (maribel4amido) [reply android/app/google-services.json:1]: Is this going to be an issue for ios part of the app as well?

File: lib/model/crop.dart [hunk 2]
```diff
@@ -0,0 +1,12 @@
+
+// TODO define UI crop model
```
Comment (maribel4amido) [inline lib/model/crop.dart:2]: For TODOs, please refer to the future JIRA user story that will complete the TODO.

File: lib/redux/home/home_state.dart [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  factory HomeState.initial() {
+    return HomeState(
+        loadingStatus: LoadingStatus.loading,
+        currentHomeTab: 0);
+  }
+
+  // Copies the states of the app or replaces for new ones if needed.
+  HomeState copyWith(
+      {LoadingStatus loadingStatus,
+      MyPlotState myPlotState,
```
Comment (maribel4amido) [inline lib/redux/home/home_state.dart:29]: Is it okay for myPlotState to be unused?

File: lib/utils/strings.dart [hunk 4]
```diff
@@ -0,0 +1,17 @@
+/// Flutter officially promotes the use of a constant class for strings.
```
Comment (maribel4amido) [inline lib/utils/strings.dart:1]: I can't see this working nicely with localisation or multi language though.

File: lib/ui/myplot/myplot_viewmodel.dart [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+  final Function fetchCrops;
+  final Function(CropEntity cropData) goToDetail;
+
+  MyPlotViewModel({this.loadingStatus,
+    this.cropsList,
+    this.fetchCrops,
+    this.goToDetail});
+
+  static MyPlotViewModel fromStore(Store<AppState> store) {
```
Comment (maribel4amido) [inline lib/ui/myplot/myplot_viewmodel.dart:19]: Can we have a todo for the goToDetail function?

File: lib/ui/myplot/my_plot_child_item.dart [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                  borderRadius: BorderRadius.circular(8.0),
+                  boxShadow: BoxShadows.plotListItemShadow()
+                ),
+                child: Column(
+                    children: <Widget>[
+                  ClipRRect(
+                    borderRadius: BorderRadius.only(
+                      topLeft: Radius.circular(8.0),
+                      topRight: Radius.circular(8.0),),
+//                      child: FadeInImage.assetNetwork(image: cropsData.imagePath, placeholder: 'assets/raw/crop_image_mock.png', fit: BoxFit.cover),
```
Comment (maribel4amido) [inline lib/ui/myplot/my_plot_child_item.dart:32]: I suggest removing any unused commented code unless it's for future work. If it is for future work, please add a TODO with a JIRA reference.

File: lib/ui/myplot/my_plot_child_item.dart [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    return GestureDetector(
+        onTap: () {
+          goToDetail(cropsData);
+        },
+        child: Padding(
+            key: ValueKey(cropsData.name ?? "No Title"),
+            padding: Margins.boxPadding(),
+            child: Container(
+                decoration: BoxDecoration(
+                  color: Colors.white,
```
Comment (maribel4amido) [inline lib/ui/myplot/my_plot_child_item.dart:22]: should all color refer to the color.dart file?

File: lib/ui/myplot/my_plot_child.dart [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+  Widget _buildBody(BuildContext context, MyPlotViewModel viewModel) {
+    switch (viewModel.loadingStatus) {
+      case LoadingStatus.loading:
+        return Container(
+            child: CircularProgressIndicator(), alignment: Alignment.center);
+      case LoadingStatus.success:
+        return _buildList(context, viewModel.cropsList, viewModel.goToDetail); // Change fetch crops to go to detail
+      case LoadingStatus.error:
+        return Text("Error");
```
Comment (maribel4amido) [inline lib/ui/myplot/my_plot_child.dart:36]: I suggest confirming with Harry what the behaviour should be when a crop cannot be loaded. A generic error my not be very useful for the user.

File: lib/ui/myplot/my_plot_child.dart [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      case LoadingStatus.x:
```
Comment (maribel4amido) [inline lib/ui/myplot/my_plot_child.dart:37]: What is the x status for?

File: lib/ui/home.dart [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    // TODO add here the 4 screens of the home
+  ];
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      body: LayoutBuilder(
+          builder: (context, constraints) => SafeArea(
+                child: Container(
+                  color: Colors.black,
```
Comment (maribel4amido) [inline lib/ui/home.dart:40]: Color reference

File: lib/main.dart [hunk 11]
```diff
@@ -1,59 +1,54 @@
...
-            }),
-        tooltip: 'Increment Counter',
-        child: Icon(Icons.add),
+    return StoreProvider<AppState>(
+      store: widget.store,
+      child: MaterialApp(
+          theme: ThemeData(
+            brightness: Brightness.light,
+            scaffoldBackgroundColor: Color(backgroundColor),
+            // ignore: strong_mode_invalid_cast_new_expr
```
Comment (maribel4amido) [inline lib/main.dart:42]: are these unused comments to be removed?

**Review by maribel4amido — CHANGES_REQUESTED** [review]

File: android/app/google-services.json (hunk 1 above)
Comment (emilianowearemobile) [reply android/app/google-services.json:1]: I did it but in my current feature branch.
I will also retro-do-it here.
iOS has its own plist, which i already added in the ignore ;)

File: lib/utils/strings.dart (hunk 4 above)
Comment (emilianowearemobile) [reply lib/utils/strings.dart:1]: Unluckily localisation is not as straight forward as in android.
https://flutter.dev/docs/development/accessibility-and-localization/internationalization

When we add zwahili support, we will have to update the way we access Strings.

File: lib/model/crop.dart (hunk 2 above)
Comment (emilianowearemobile) [reply lib/model/crop.dart:2]: I agree.
Still, i prefer to delete it until we need it for real. 
I will delete it in the current feature branch.

File: lib/redux/home/home_state.dart (hunk 3 above)
Comment (emilianowearemobile) [reply lib/redux/home/home_state.dart:29]: Already deleted it in the detail branch :)

File: lib/ui/myplot/myplot_viewmodel.dart (hunk 5 above)
Comment (emilianowearemobile) [reply lib/ui/myplot/myplot_viewmodel.dart:19]: Its already functioning in the current feature build :)

File: lib/ui/myplot/my_plot_child_item.dart (hunk 6 above)
Comment (emilianowearemobile) [reply lib/ui/myplot/my_plot_child_item.dart:32]: Its already functioning in the current feature build :)

File: lib/ui/myplot/my_plot_child_item.dart (hunk 7 above)
Comment (emilianowearemobile) [reply lib/ui/myplot/my_plot_child_item.dart:22]: We should reference all colors to colors.dart and access them Colors(white).
Added the change in next builds.

File: lib/ui/myplot/my_plot_child.dart (hunk 8 above)
Comment (emilianowearemobile) [reply lib/ui/myplot/my_plot_child.dart:36]: Clearly not, what's missing here is a TODO.
I will add it + a ticket.

File: lib/ui/myplot/my_plot_child.dart (hunk 9 above)
Comment (emilianowearemobile) [reply lib/ui/myplot/my_plot_child.dart:37]: I thought of using it as noPlots (and updating the name to reflect this).
But i will manage it differently, by asking how many plots has the user after the request.

File: lib/data/managers/firestore_manager.dart [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  FireStoreManager._internal();
+
+  Future<List<CropEntity>> getCrops() async {
+    List<CropEntity> cropsEntities;
+
+    // Filters defined by product definition.
+    var query = Firestore.instance
+        .collection('fl_content')
```
Comment (albertmiro) [inline lib/data/managers/firestore_manager.dart:20]: Would be nice if we can move these values to constants

File: lib/data/managers/firestore_manager.dart [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        crop.setImagePath(getImagePath(imageSnapshot));
+        cropsEntitiesWithImagePath.add(crop);
+      });
+    }
+    return cropsEntitiesWithImagePath;
+  }
+
+
+  String getImagePath(DocumentSnapshot imageDocument) {
+    final sizePath = imageDocument.data["sizes"].first["path"];
```
Comment (albertmiro) [inline lib/data/managers/firestore_manager.dart:52]: Not sure about how that works in Flutter, but if the value of data is null, that can give you a Null Poiner?

File: lib/data/managers/firestore_manager.dart [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    final imageFileNamePath = imageDocument.data["file"];
+    return IMAGE_BASE_PATH + "/" + sizePath + "/" + imageFileNamePath;
```
Comment (albertmiro) [inline lib/data/managers/firestore_manager.dart:54]: Move "/" as a const value

File: lib/data/model/crop_entity.dart [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    imagePath: "",
+    name: cropDocument.data["name"],
+    nonCompanionPlants: cropDocument.data["nonCompanionPlants"] == null ? null : new List<String>.from(cropDocument.data["nonCompanionPlants"].map((x) => x)),
+    profitability: cropDocument.data["profitability"] == null ? null : loHiValues.map[cropDocument.data["profitability"]],
+    setupCost: cropDocument.data["setupCost"] == null ? null : loHiValues.map[cropDocument.data["setupCost"]],
+    soilType: cropDocument.data["soilType"] == null ? null : new List<String>.from(cropDocument.data["soilType"].map((x) => x)),
+    stages: "",
+    status: statusValues.map[cropDocument.data["status"]],
+    summary: cropDocument.data["summary"],
+    waterRequirement: cropDocument.data["waterRequirement"] == null ? null : loHiValues.map[cropDocument.data["waterRequirement"]],
```
Comment (albertmiro) [inline lib/data/model/crop_entity.dart:59]: Would be nice if we can refactor that CropEntity object creation maybe creating some methods to do not have this logic when assigning the value

File: lib/model/loading_status.dart [hunk 16]
```diff
@@ -0,0 +1,6 @@
+enum LoadingStatus {
+  loading,
+  error,
+  success,
+  x
```
Comment (albertmiro) [inline lib/model/loading_status.dart:5]: Can we have the enum values in Mayus?

File: lib/redux/app/app_reducer.dart [hunk 17]
```diff
@@ -0,0 +1,9 @@
+import 'package:farmsmart_flutter/redux/app/app_state.dart';
+import 'package:farmsmart_flutter/redux/home/home_reducer.dart';
+import 'package:farmsmart_flutter/redux/home/myPlot/my_plot_reducer.dart';
+
+AppState appReducer(AppState state, dynamic action) =>
+    new AppState(
```
Comment (albertmiro) [inline lib/redux/app/app_reducer.dart:6]: `new` can be removed

File: lib/model/enums.dart [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+final loHiValues = new EnumValues({
+  "HIGH": LoHi.HIGH,
+  "LOW": LoHi.LOW,
+  "MEDIUM": LoHi.MEDIUM
+});
+
+enum Status { DRAFT, PUBLISHED }
+
+final statusValues = new EnumValues({
```
Comment (albertmiro) [inline lib/model/enums.dart:28]: `new` can be removed

File: lib/model/enums.dart [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+enum CropType { SINGLE, ROTATION }
+
+final cropTypeValues = new EnumValues({
+  "ROTATION": CropType.ROTATION,
+  "SINGLE": CropType.SINGLE
+});
+
+enum LoHi { LOW, MEDIUM, HIGH }
+
+final loHiValues = new EnumValues({
```
Comment (albertmiro) [inline lib/model/enums.dart:20]: `new` can be removed

File: lib/model/enums.dart [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+
+final begAdvValues = new EnumValues({
+  "ADVANCED": CropComplexity.ADVANCED,
+  "BEGINNER": CropComplexity.BEGINNER,
+  "INTERMEDIATE": CropComplexity.INTERMEDIATE
+});
+
+enum CropType { SINGLE, ROTATION }
+
+final cropTypeValues = new EnumValues({
```
Comment (albertmiro) [inline lib/model/enums.dart:13]: `new` can be removed

File: lib/model/enums.dart [hunk 21]
```diff
@@ -0,0 +1,45 @@
+
+
+enum CropComplexity { BEGINNER, INTERMEDIATE, ADVANCED }
+
+final begAdvValues = new EnumValues({
```
Comment (albertmiro) [inline lib/model/enums.dart:5]: `new` can be removed

File: lib/redux/home/home_reducer.dart [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+    case HomeScreen.PROFIT_LOSS_TAB:
+      return state.copyWith(currentHomeTab: 1);
+      break;
+    case HomeScreen.ARTICLES_TAB:
+      return state.copyWith(currentHomeTab: 2);
+      break;
+    case HomeScreen.COMMUNITY_TAB:
+      return state.copyWith(currentHomeTab: 3);
+      break;
+  }
```
Comment (albertmiro) [inline lib/redux/home/home_reducer.dart:26]: Can we move this tabs positions values, to constants?
Can we have defined something like:

```
const MY_PLOT_TAB = 0
const PROFIT_LOSS_TAB = 1
const ARTICLES_TAB = 2
const COMMUNITY_TAB = 3

```
And then use them wherever we need to access at these positions, in this way, if the tabs change their position we don't have to go around and check where is the number we set.

File: lib/redux/home/home_state.dart [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+
+  HomeState(
+      {@required this.loadingStatus,
+      @required this.currentHomeTab});
+
+  // We set all the states initial values
+  factory HomeState.initial() {
+    return HomeState(
+        loadingStatus: LoadingStatus.loading,
+        currentHomeTab: 0);
```
Comment (albertmiro) [inline lib/redux/home/home_state.dart:23]: Here is one case where we would be able to use the const with the tab number

File: lib/ui/app_bar.dart [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+      case 0: // MyPlot
+        return buildForHome(Strings.myPlotTab, profileAction(), homeIcon() );
+        break;
+      case 1: // Profit / Loss
+        return buildWithTitle(Strings.profitLossTab);
+        break;
+      case 2: // Articles
+        return buildWithTitle(Strings.discoverTab);
+        break;
+      case 3: // Community
```
Comment (albertmiro) [inline lib/ui/app_bar.dart:21]: Here is another place where we would be able to use the const values for the tabs

File: lib/ui/home.dart [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+          child: BottomNavigationBar(
+            type: BottomNavigationBarType.fixed,
+            onTap: viewModel.changeTab,
+            currentIndex: viewModel.currentTab,
+            items: [
+              BottomNavigationBarItem(
+                activeIcon: Image.asset("assets/icons/my_plot_selected.png", height: bottomBarIconSize),
+                icon: Image.asset("assets/icons/my_plot.png", height: bottomBarIconSize),
+                title: Text(Strings.myPlotTab),
+              ),
```
Comment (albertmiro) [inline lib/ui/home.dart:66]: Can we create a method than returns to us a Widget where we send as enter parameters the Active icon, the icon, and the title. In this way we can use that wherever we need ;-)

File: lib/ui/home_viewmodel.dart [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+      case 0:
+        showMyPlotChild();
+        break;
+      case 1:
+        showProfitLossChild();
+        break;
+      case 2:
+        showArticlesChild();
+        break;
+      case 3:
```
Comment (albertmiro) [inline lib/ui/home_viewmodel.dart:48]: Here we could use the const for the tabs too

File: lib/ui/profitloss/profit_loss_child.dart [hunk 27]
```diff
@@ -0,0 +1,13 @@
+import 'package:flutter/material.dart';
+
+class HomeProfitLossChild extends StatelessWidget {
+  HomeProfitLossChild();
+
+  @override
+  Widget build(BuildContext context) {
+    return SizedBox.expand(
+      child: Image.asset('assets/raw/profit_mock.png', fit: BoxFit.cover),
```
Comment (albertmiro) [inline lib/ui/profitloss/profit_loss_child.dart:9]: Move that to a const value :-)

File: lib/ui/myplot/my_plot_child.dart (hunk 8 above)
Comment (emilianowearemobile) [reply lib/ui/myplot/my_plot_child.dart:36]: Created FARM-203

File: lib/ui/myplot/my_plot_child.dart (hunk 9 above)
Comment (emilianowearemobile) [reply lib/ui/myplot/my_plot_child.dart:37]: Deleted this state :)

File: lib/main.dart (hunk 11 above)
Comment (emilianowearemobile) [reply lib/main.dart:42]: Removed :)

File: lib/data/managers/firestore_manager.dart (hunk 12 above)
Comment (emilianowearemobile) [reply lib/data/managers/firestore_manager.dart:20]: Agree!

File: lib/data/managers/firestore_manager.dart (hunk 13 above)
Comment (emilianowearemobile) [reply lib/data/managers/firestore_manager.dart:52]: Well, as this is a flamelink document reference, should point to a non-null document.
But still we can both null check or try - catch the whole operation 👎what do you think?

File: lib/model/loading_status.dart (hunk 16 above)
Comment (emilianowearemobile) [reply lib/model/loading_status.dart:5]: Sure!

File: lib/data/managers/firestore_manager.dart (hunk 14 above)
Comment (emilianowearemobile) [reply lib/data/managers/firestore_manager.dart:54]: No actual need for this...
this reads better than  IMAGE_BASE_PATH + DASH + sizePath + DASH+ imageFileNamePath

File: lib/data/model/crop_entity.dart (hunk 15 above)
Comment (emilianowearemobile) [reply lib/data/model/crop_entity.dart:59]: I will refactor some setters to make it more readable!

File: lib/ui/home.dart (hunk 25 above)
Comment (emilianowearemobile) [reply lib/ui/home.dart:66]: It's doable but i think a first better step would be to change the assets into constants.

File: lib/ui/profitloss/profit_loss_child.dart (hunk 27 above)
Comment (emilianowearemobile) [reply lib/ui/profitloss/profit_loss_child.dart:9]: Done 👨‍🚀

File: lib/data/managers/firestore_manager.dart (hunk 13 above)
Comment (albertmiro) [reply lib/data/managers/firestore_manager.dart:52]: Maybe a null check is ok

File: lib/model/crop.dart (hunk 2 above)
Comment (maribel4amido) [reply lib/model/crop.dart:2]: fair enough

File: android/app/google-services.json (hunk 1 above)
Comment (maribel4amido) [reply android/app/google-services.json:1]: Thanks for that

File: lib/utils/strings.dart (hunk 4 above)
Comment (maribel4amido) [reply lib/utils/strings.dart:1]: I think the action here is to create a JIRA to address localisation. Ideally, the localisation framework should be in place even if we still only have a single language.

File: lib/ui/myplot/my_plot_child.dart (hunk 8 above)
Comment (maribel4amido) [reply lib/ui/myplot/my_plot_child.dart:36]: Thanks!

**Review by maribel4amido — APPROVED** [review]

File: lib/utils/strings.dart (hunk 4 above)
Comment (emilianowearemobile) [reply lib/utils/strings.dart:1]: As you please. I think its part of the zwahili set of tasks, so no need to do it before hand.

File: lib/data/managers/firestore_manager.dart (hunk 13 above)
Comment (emilianowearemobile) [reply lib/data/managers/firestore_manager.dart:52]: Done

**Pull request merged**